### PR TITLE
Project cloud-storage upload

### PR DESCRIPTION
**Pull request by Kukdo**
Project cloud-storage uploads ver 0.1 & 0.2

### CODE REVIEW

File: jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/model/FileMetadata.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+
+import java.time.LocalDateTime;
+
+@Entity
+public class FileMetadata {
```
Comment (xiacongling) [inline jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/model/FileMetadata.java:11]: in mvc, the "model" represents the data and the business logic of the application

File: jzp-code/kukdo/P2288_0618.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
+public class Solution {
```
Comment (xiacongling) [inline jzp-code/kukdo/P2288_0618.java:1]: move the exercises to a another PR

File: jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/service/impl/AliyunStorageServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    @Value("${aliyun.oss.bucketName}")
+    private String bucketName;
+
+    private OSS getOSSClient() {
+        ClientConfiguration conf = new ClientConfiguration();
+//        conf.setConnectionTimeout(5000);
+//        conf.setSocketTimeout(10000);
+//        conf.setMaxErrorRetry(3);
+        return new OSSClientBuilder().build(endpoint, accessKeyId, accesKeySecret);
+    }
```
Comment (xiacongling) [inline jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/service/impl/AliyunStorageServiceImpl.java:37]: we need not creating oss clients every time we use it. it can be a singleton and its lifecycle should be the same as that of the application.

File: jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/service/impl/AliyunStorageServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    private String endpoint;
+
+    @Value("${aliyun.oss.accessKeyId}")
+    private String accessKeyId;
+
+    @Value("${aliyun.oss.accessKeySecret}")
+    private String accesKeySecret;
+
+    @Value("${aliyun.oss.bucketName}")
+    private String bucketName;
```
Comment (xiacongling) [inline jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/service/impl/AliyunStorageServiceImpl.java:29]: consider `@ConfigurationProperties`

File: jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/service/impl/LocalStorageServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+    @Override
+    public void delete(String filePath) {
+
+    }
+
+    @Override
+    public void rename(String oldFileName, String newFileName) {
+
+    }
```
Comment (xiacongling) [inline jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/service/impl/LocalStorageServiceImpl.java:55]: not implemented?

**Review by xiacongling — CHANGES_REQUESTED** [review]

File: jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/service/impl/LocalStorageServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+@Service("LocalStorage")
+public class LocalStorageServiceImpl implements StorageService {
+
+    private final Path rootLocation = Paths.get("upload_file");
```
Comment (xiacongling) [inline jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/service/impl/LocalStorageServiceImpl.java:17]: should be configurable

File: jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/service/impl/LocalStorageServiceImpl.java (hunk 5 above)
Comment (Kukdo) [reply jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/service/impl/LocalStorageServiceImpl.java:55]: > not implemented?

yes, not yet until this afternoon.

File: jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/service/impl/LocalStorageServiceImpl.java (hunk 6 above)
Comment (Kukdo) [reply jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/service/impl/LocalStorageServiceImpl.java:17]: > should be configurable

yes, here just for test yesterday, I will add a property in `application.properties` and update this part.

File: jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/service/impl/AliyunStorageServiceImpl.java (hunk 4 above)
Comment (Kukdo) [reply jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/service/impl/AliyunStorageServiceImpl.java:29]: > consider `@ConfigurationProperties`

Thanks for the suggestion, I created a new class `AliyunProperties` for storing Aliyun configurations, thus separating the configuration properties from the business logic, using @ConfigurationProperties(prefix = "aliyun.oss") and @EnableConfigurationProperties( AliyunProperties.class).

File: jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/service/impl/AliyunStorageServiceImpl.java (hunk 3 above)
Comment (Kukdo) [reply jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/service/impl/AliyunStorageServiceImpl.java:37]: > we need not creating oss clients every time we use it. it can be a singleton and its lifecycle should be the same as that of the application.

Thanks for advice, I created a new `OssClientConfig` configuration class to manage instances of OSSClient. This class will configure the OSSClient as a Spring bean so that it remains singleton throughout the application lifecycle.

File: jzp-code/kukdo/P2288_0618.java (hunk 2 above)
Comment (Kukdo) [reply jzp-code/kukdo/P2288_0618.java:1]: > move the exercises to a another PR

working on it.

Comment (Kukdo) [reply jzp-code/kukdo/P2288_0618.java]: finished.

File: jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/pom.xml [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    </parent>
+    <groupId>com.xiaomi</groupId>
+    <artifactId>cloud-storage-parent</artifactId>
+    <version>0.0.1-SNAPSHOT</version>
+    <packaging>pom</packaging>
+    <modules>
+        <module>./cloud-storage-web</module>
+    </modules>
+    <properties>
+        <java.version>17</java.version>
```
Comment (xiacongling) [inline jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/pom.xml:19]: do we need `maven.compiler.release`?

File: jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/controller/FileController.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    @PostMapping("/files")
+    public ResponseEntity<FileMetadata> uploadFile(@RequestParam("file") MultipartFile file) {
+        FileMetadata metadata = fileService.storeFile(file);
+        return ResponseEntity.ok(metadata);
+    }
+
+    @GetMapping("/files/{Id}")
+    public ResponseEntity<Resource> downloadFile(@PathVariable Long Id) {
```
Comment (xiacongling) [inline jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/controller/FileController.java:40]: argument name (`Id`) should start with a lowercase letter

File: jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/controller/FileController.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+            } else {
+                throw new RuntimeException("File not found or not readable" + metadata.getFilePath());
+            }
+        } catch (Exception e) {
+            throw new RuntimeException("Failed to download file: " + e.getMessage());
+        }
+    }
+
+    @DeleteMapping("/files/{Id}")
+    public ResponseEntity<ApiResponse> deleteFile(@PathVariable long Id) {
```
Comment (xiacongling) [inline jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/controller/FileController.java:57]: `id`

File: jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/controller/FileController.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+        fileService.deleteFile(Id);
+        return ResponseEntity.ok(new ApiResponse(true, "File deleted successfully"));
+    }
+
+    @PutMapping("/files/{Id}")
+    public ResponseEntity<ApiResponse> renameFile(@PathVariable long Id, @RequestParam("newName") String newName) {
```
Comment (xiacongling) [inline jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/controller/FileController.java:63]: `id`

File: jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/controller/FileController.java [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+        fileService.renameFile(Id, newName);
```
Comment (xiacongling) [inline jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/controller/FileController.java:64]: rename? how can I update the file content?

File: jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/unfin/CloudStorageClient.java [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+
+@Command(name = "cloudstorage", mixinStandardHelpOptions = true, version = "1.0",
+        description = "Client for interacting with cloud storage service")
+public class CloudStorageClient implements Runnable {
+
+    @Option(names = {"-u", "--upload"}, description = "Upload file to cloud storage")
```
Comment (xiacongling) [inline jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/unfin/CloudStorageClient.java:25]: use a sub-command is better, like:
```
my-cloud-storage upload --name <file-name> --file <client-file-path>
my-cloud-storage download --name <file-name> --destination <client-file-path>
my-cloud-storage delete --name <file-name>
my-cloud-storage list
```

File: jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/unfin/CloudStorageClient.java [hunk 13]
```diff
@@ -0,0 +1,124 @@
+package com.xiaomi.unfin;
```
Comment (xiacongling) [inline jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/unfin/CloudStorageClient.java:1]: can we make a new pr for this functionality?

**Review by xiacongling — COMMENTED** [review]
can you add an `application.properties` or `appication.yaml` for development environment?

File: jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/controller/FileController.java (hunk 8 above)
Comment (Kukdo) [reply jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/controller/FileController.java:40]: > argument name (`Id`) should start with a lowercase letter
got it, I will replace all the `Id` with `id` in the later version.

File: jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/pom.xml (hunk 7 above)
Comment (Kukdo) [reply jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/pom.xml:19]: > do we need `maven.compiler.release`?
yes, thanks for pointing it out.

**Comment by Kukdo** [comment]
> can you add an `application.properties` or `appication.yaml` for development environment?

I'll add `application.properties` in the next version, after wiping out critical private information of cloud service.

File: jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/unfin/CloudStorageClient.java (hunk 13 above)
Comment (Kukdo) [reply jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/unfin/CloudStorageClient.java:1]: > can we make a new pr for this functionality?

of course.

File: jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/controller/FileController.java (hunk 11 above)
Comment (Kukdo) [reply jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/controller/FileController.java:64]: > rename? how can I update the file content?

For the cloud drive, I think updating the file content is the equivalent of uploading a new file, whose corresponding id should be different from the previous one.

File: jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/unfin/CloudStorageClient.java (hunk 12 above)
Comment (Kukdo) [reply jzp-code/kukdo/project/cloud-storage/cloud-storage-parent/cloud-storage-web/src/main/java/com/xiaomi/unfin/CloudStorageClient.java:25]: > use a sub-command is better, like:
> 
> ```
> my-cloud-storage upload --name <file-name> --file <client-file-path>
> my-cloud-storage download --name <file-name> --destination <client-file-path>
> my-cloud-storage delete --name <file-name>
> my-cloud-storage list
> ```

I'll fix the bugs of this part and implement it based on your productive suggestions, the new PR is expected on 6.25

**Review by xiacongling — APPROVED** [review]
